File: jetbrains-plugin/src/main/kotlin/com/ezevillo/gitreview/domain/NameStatus.kt
package com.ezevillo.gitreview.domain

/**
 * A file touched by a commit, with resolved before/after paths.
 * `null` means the file does not exist on that side.
 */
data class CommitChange(
    val path: String,
    val before: String?,
    val after: String?,
)

/**
 * Parses `git diff-tree -z --name-status` / `git diff --name-status -z` output.
 * With `-z`, git never quotes paths.
 */
fun parseNameStatus(output: String): List<CommitChange> {
    val fields = output.split('\u0000').filter { it.isNotEmpty() }
    val changes = ArrayList<CommitChange>()
    var i = 0
    while (i < fields.size) {
        val code = fields[i].firstOrNull() ?: break
        if (code == 'R' || code == 'C') {
            val from = fields.getOrNull(i + 1) ?: break
            val to = fields.getOrNull(i + 2) ?: break
            changes.add(CommitChange(path = to, before = from, after = to))
            i += 3
            continue
        }
        val path = fields.getOrNull(i + 1) ?: break
        changes.add(
            CommitChange(
                path = path,
                before = if (code == 'A') null else path,
                after = if (code == 'D') null else path,
            ),
        )
        i += 2
    }
    return changes
}

File: visualstudio-extension/src/GitReview.Domain/NameStatus.cs
namespace GitReview.Domain;

/// <summary>
/// A file touched by a commit, with resolved before/after paths.
/// null means the file does not exist on that side.
/// </summary>
public sealed record CommitChange(string Path, string? Before, string? After);

public static class NameStatus
{
    /// <summary>
    /// Parses <c>git diff-tree -z --name-status</c> / <c>git diff --name-status -z</c> output.
    /// With -z, git never quotes paths.
    /// </summary>
    public static IReadOnlyList<CommitChange> ParseNameStatus(string output)
    {
        var fields = output.Split('\0').Where(f => f.Length > 0).ToList();
        var changes = new List<CommitChange>();
        var i = 0;
        while (i < fields.Count)
        {
            var code = fields[i].Length > 0 ? fields[i][0] : '\0';
            if (code is 'R' or 'C')
            {
                if (i + 2 >= fields.Count) break;
                var from = fields[i + 1];
                var to = fields[i + 2];
                changes.Add(new CommitChange(to, from, to));
                i += 3;
                continue;
            }
            if (i + 1 >= fields.Count) break;
            var path = fields[i + 1];
            changes.Add(new CommitChange(
                path,
                Before: code == 'A' ? null : path,
                After: code == 'D' ? null : path));
            i += 2;
        }
        return changes;
    }
}
